Refactor `_convert_container`

Charlie-XIAO opened the pull request:
Follow up of #28521, see https://github.com/scikit-learn/scikit-learn/pull/28521#discussion_r1507741366.

This PR splits `constructor_name` into `constructor_type` and `constructor_lib`, and adds two parameters `sparse_container` and `sparse_format` to reflect dense versus sparse. /cc @glemaitre

#### Mapping from the original `constructor_name` to the specifications in this PR

|          | `constructor_type` | `constructor_lib` | `sparse_container` | `sparse_format` |
| :------: | :----------------: | :---------------: | :----------------: | :-------------: |
| Default  | | `"pandas"`        | `None`             | `"csr"`         |
| `"list"` | `"list"` | | | |
| `"tuple"` | `"tuple"` | | | |
| `"slice"` | `"slice"` | | | |
| `"array"` | `"array"` | | | |
| `"sparse"`/`"sparse_csr"` | `"array"` | | `"matrix"` | |
| `"sparse_csc"` | `"array"` | | `"matrix"` | `"csc"` |
| `"sparse_csr_array"` | `"array"` | | `"array"` | |
| `"sparse_csc_array"` | `"array"` | | `"array"` | `"csc"` |
| `"dataframe"`/`"pandas"` | `"dataframe"` | | | |
| `"polars"` | `"dataframe"` | `"polars"` | | |
| `"pyarrow"` | `"dataframe"` | `"pyarrow"` | | |
| `"series"` | `"series"` | | | |
| `"polars_series"` | `"series"` | `"polars"` | | |
| `"pyarrow_array"` | `"series"` | `"pyarrow"` | | |
| `"index"` | `"index"` | | | |

#### Some rules of when each parameter is applicable

- `sparse_container` has no effect unless `constructor_type="array"`
- `sparse_format` has no effect unless `constructor_type="array"` and `sparse_container` is not `None`
- `constructor_lib` has no effect unless `constructor_type` is one of `"dataframe"`, `"series"`, and `"index"`
- `minversion` has no effect unless `constructor_lib` is used
- `column_names` (renamed from `columns_name`) has no effect unless `constructor_type="dataframe"`
- `categorical_feature_names` has no effect unless `constructor_type="dataframe"`

#### Constraints on the shape of the input container

- `constructor_type="slice"` requires 1-dimensional container with exactly 2 elements
- `constructor_type="dataframe"` requires 2-dimensional container (previously it converts `(n,)` shape to `(n, 1)` shape)
- `constructor_type="series"` requires 1-dimensional container (previously it creates series with each element being a list when given 2-dimensional container)
- `constructor_type="index"` requires 1-dimensional container (similar to `"series"`)

Charlie-XIAO committed:
RFC logic of _convert_container

Charlie-XIAO committed:
update more tests

Charlie-XIAO committed:
update more tests

Charlie-XIAO committed:
update more tests

Charlie-XIAO committed:
update more tests, except test_testing and test_indexing

Charlie-XIAO committed:
fix tests for _indexing

Charlie-XIAO committed:
fix tests for _testing

Charlie-XIAO committed:
refactor to add more error messages

Charlie-XIAO committed:
fix error in test_indexing

Charlie-XIAO committed:
improve coverage

Charlie-XIAO committed:
add changelog to reflect a fix for _safe_indexing

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -747,99 +750,166 @@ def _convert_container(
-            raise ValueError(
-                f"{constructor_name} is only available with scipy>=1.8.0, got "
-                f"{sp_version}"
+        if sparse_container == "array":
+            sparse_container = sp.sparse.csr_array(container)
+        else:
+            sparse_container = sp.sparse.csr_matrix(container)
+        return sparse_container.asformat(sparse_format)
```
We can avoid the double conversion

jeremiedbb replied:
I was just thinking using ``if`` clauses since we only have 2 options :)

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -747,99 +750,166 @@ def _convert_container(
+        Force the dtype of the container. Does not apply to "slice".
+
+    sparse_container : {"matrix", "array"}, default=None
+        The sparse container to use. Only applies to "array". Note that if this
```
I don't understand ``Only applies to "array"``.  array here refers to the constructor type ? Maybe
```suggestion
        The sparse container to use. Only applies when `constructor_type="array"`. Note that if this
```

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -747,99 +750,166 @@ def _convert_container(
+        - "array" returns a sparse array and skip test if scipy < 1.8
+
+    sparse_format : {"csc", "csr"}, default="csr"
+        The sparse format to use. Only applies to "array" and when `sparse_container`
```
```suggestion
        The sparse format to use. Only applies when `constructor_type="array"` and when `sparse_container`
```

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -747,99 +750,166 @@ def _convert_container(
+                f"Only 1D containers can be converted to {constructor_type}; got shape "
+                f"{container.shape} instead"
+            )
+
+    # ---- Conversion ---------------------------------------------------------
+
```
We usually don't use that kind of comment style. Since the function is getting pretty big, I think it could make sense to extract the individual converters for complex constructor types into dedicated smaller functions

Charlie-XIAO replied:
Thanks for the suggestion. I've created separate functions `_convert_container_to_{constructor_type}` for `"array"`, `"dataframe"`, `"series"`, and `"index"`. Those one-liners are left as is.

jeremiedbb commented on sklearn/utils/tests/test_indexing.py:
```diff
@@ -28,6 +29,62 @@
+                    "constructor_lib": series_lib,
+                }
+        else:
+            raise ValueError(f"Invalid constructor type: {constructor_type}")  # noqa
```
I think you meant
```suggestion
            raise ValueError(f"Invalid constructor type: {constructor_type}")  # pragma: no cover
```

Charlie-XIAO replied:
Thanks for pointing out!

jeremiedbb reviewed (commented):
Thanks for the PR @Charlie-XIAO. Here are some comments. I'm +1 for the refactoring of ``_convert_container``.

Charlie-XIAO committed:
resolve conversations and make docstring clearer

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

glemaitre commented:
Once the CIs are green, I'll review ;) (I am buying a bit of time).
But indeed, I really like the refactoring. Thanks for taking care of it @Charlie-XIAO.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
fix stupid error (forgetting to return); more errmsgs and tests

Charlie-XIAO commented:
Now I have resolved all suggestions from Jérémie, and CI is green :)

Charlie-XIAO committed:
Merge branch 'main' into rfc-convert-container

glemaitre commented:
I'm going to resolve the conflicts and make a review.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/Charlie-XIAO/28681

glemaitre commented on sklearn/utils/_testing.py:
```diff
@@ -736,111 +736,237 @@ def assert_run_python_script_without_output(source_code, pattern=".+", timeout=6
+        - "series" and "index" requires 1D container
+
+    constructor_type : {"list", "tuple", "slice", "array", "dataframe", "series", \
+                        "index"}
         The type of the returned container.
```
```suggestion
            "index"}
        The type of the returned container.
```

glemaitre commented on sklearn/utils/_testing.py:
```diff
@@ -736,111 +736,237 @@ def assert_run_python_script_without_output(source_code, pattern=".+", timeout=6
+        if dtype is not None:
+            container = container.astype(dtype, copy=False)
+    else:
+        container = np.asarray(container, dtype=dtype)
```
If I get it right, we are going to convert any dataframe to an array at this stage. I'm not sure this is wise since I could imagine converting a dataframe pandas into a dataframe polars but I would not need to go back to numpy, isn't it?

glemaitre replied:
In some way, I would as much as possible delay calling `asarray` to be sure that we actually need it.

Charlie-XIAO replied:
I didn't think that much of performance given this is a testing utility, but your point is valid. This would mean refactoring and I will do it when I get some full block of time :)

glemaitre replied:
> I didn't think that much of performance given this is a testing utility, but your point is valid. 

I'm a bit more scared about the `dtype` casting where will enforce a single `dtype` while we might be interested in dataframe with heterogeneous dtype and just trying to test for pandas and polars. Passing by NumPy in the middle could make it a mess :)

Charlie-XIAO replied:
@glemaitre I have again refactored the code, and in particular sparse-to-sparse, dataframe-to-dataframe, and series-to-series are dealt with differently. Moreover, now `dtype` should specify the `dtype` of the target (e.g., converting to polars would require a polars dtype) and directly affect the converted container, instead of specifying dtype on the array before conversion (as I previously did).

glemaitre commented on sklearn/utils/_testing.py:
```diff
@@ -736,111 +736,237 @@ def assert_run_python_script_without_output(source_code, pattern=".+", timeout=6
+            f"Invalid {sparse_format=}; expected one of {supported_sparse_formats}"
+        )
+
+    if sparse_container == "array":
+        if sparse_format == "csr":
+            container = sp.sparse.csr_array(container)
+        else:  # csc
+            container = sp.sparse.csc_array(container)
+    else:  # matrix
+        if sparse_format == "csr":
+            container = sp.sparse.csr_matrix(container)
+        else:  # csc
+            container = sp.sparse.csc_matrix(container)
```
I think we can simplify with an f-string and `getattr`.

```suggestion
    sparse_constructor = getattr(sp.sparse, f"{sparse_format}_{sparse_container}")
    container = sparse_constructor(container)
```

glemaitre reviewed (commented):
A couple of comments before to look at the tests.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
partial resolution of suggestions; refactoring will be done later

Charlie-XIAO committed:
Merge branch 'main' into rfc-convert-container

Charlie-XIAO committed:
fix linting

Charlie-XIAO committed:
fix tests

Charlie-XIAO committed:
convert to numpy array only when really needed

Charlie-XIAO committed:
code coverage and fix pyarrow

Charlie-XIAO committed:
move changelog entry

Charlie-XIAO committed:
remove unexpected changes

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
test coverage series to series

Charlie-XIAO committed:
pragma no cover

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
simplify a bit for categorical polars

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

adrinjalali commented:
@Charlie-XIAO would you mind resolving conflicts?

Charlie-XIAO commented:
Will do soon (maybe within one day or two).

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
fix changelog

Charlie-XIAO committed:
fix changelog (skip check otherwise line terminators will be modified on windows)

Charlie-XIAO commented:
Sorry for the super long delay - I was too busy with coursework during the semester. I have now resolved the conflicts, but not sure if anything has changed since my last update since it is affecting quite a lot of tests.

Charlie-XIAO committed:
fix missing pyarrow

Charlie-XIAO commented:
The CI failure seems to be because that `to_pandas` of polars requires pyarrow as well: https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.to_pandas.html, should we just skip the tests of polars when pyarrow is not installed, or try find a workaround?

adrinjalali commented on sklearn/utils/_testing.py:
```diff
@@ -925,111 +928,339 @@ def assert_run_python_script_without_output(source_code, pattern=".+", timeout=6
-    elif constructor_name == "polars_series":
+        if container_lib is not None:
+            if container_lib != "pandas":
+                container = container.to_pandas()
```
I vaguely remember @MarcoGorelli mentioning an arrow-free path for polars->pandas, or am I misremembering?

MarcoGorelli replied:
hey - if you have numeric data without missing values (or if you don't care about null vs nan) you could do
```python
In [1]: import pandas as pd

In [2]: import polars as pl

In [3]: df = pl.DataFrame({'a': [1,2,3], 'b': [4,5,6]})

In [4]: pd.DataFrame({key: df[key] for key in df.columns})
Out[4]:
   a  b
0  1  4
1  2  5
2  3  6

In [5]: import pyarrow
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[5], line 1
----> 1 import pyarrow

ModuleNotFoundError: No module named 'pyarrow'
```

Charlie-XIAO replied:
Thanks! Fixed now

Charlie-XIAO committed:
use arrow-free polars-to-pandas conversion suggested by MarcoGorelli

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -2365,6 +2365,15 @@ def _is_polars_df(X):
     return isinstance(X, pl.DataFrame)
 
 
+def _is_arrow_df(X):
```
do we have a CI with arrow installed? (I think we should)

Charlie-XIAO replied:
Yes I think we do have it in `pylatest_conda_forge_mkl_linux-64`:

https://github.com/scikit-learn/scikit-learn/blob/c9aeb15f8f1c7c54ed4ef27c871f7167e2ce3077/build_tools/azure/pylatest_conda_forge_mkl_linux-64_environment.yml#L29

adrinjalali reviewed (approved):
Nice work, thanks @Charlie-XIAO .

Charlie-XIAO committed:
Merge branch 'main' into rfc-convert-container

Charlie-XIAO commented:
Will do tomorrow

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO commented:
This has changed too much since I last updated. I need to take some time to check the new pyarrow-related changes to make sure I did not mistakenly drop any test, thus marking as draft now.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
modify the scheme to put pyarrow array in

Charlie-XIAO committed:
Revert "modify the scheme to put pyarrow array in"

This reverts commit 11a6186dae1b84a19b10357b4b4fc44c2489dc70.

Charlie-XIAO committed:
add pyarrow (not completed)

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into rfc-convert-container

Charlie-XIAO committed:
fix test_convert_container

Charlie-XIAO committed:
add pyarrow dataframe (table) and series (array) to testing

Charlie-XIAO commented:
@adrinjalali I've added "pyarrow_array" as `{"constructor_type": "series", "constructor_lib": "pyarrow"` and added (at least) the additional tests in #31040. This PR is ready for review now.

adrinjalali reviewed (commented):
This LGTM, and we should probably merge quick not to have more conflicts.

adrinjalali commented:
@jeremiedbb maybe you could have another look?
